src/main/kotlin/io/libp2p/core/multiformats/Protocol.kt

```diff
@@ -166,3 +103,100 @@ enum class Protocol(val code: Int, val size: Int, val typeName: String) {
         fun getOrThrow(name: String) = get(name) ?: throw IllegalArgumentException("Unknown protocol name: '$name'")
     }
 }
+
+private const val LENGTH_PREFIXED_VAR_SIZE = -1
+
+private val SIZE_VALIDATOR: (Protocol, ByteArray?) -> Unit = { protocol, bytes ->
+    if (!protocol.hasValue && bytes != null)
+        throw IllegalArgumentException("No value expected for protocol $protocol, but got ${bytes.contentToString()}")
+    if (protocol.hasValue) {
+        requireNotNull(bytes) { "Non-null value expected for protocol $protocol" }
+        if (protocol.sizeBits != LENGTH_PREFIXED_VAR_SIZE && bytes.size * 8 != protocol.sizeBits) {
+            throw IllegalArgumentException("Value of size ${protocol.sizeBits / 8} expected for protocol $protocol but got ${bytes.contentToString()}")
+        }
+    }
+}
+private val NON_EMPTY_VALIDATOR: (Protocol, ByteArray?) -> Unit = { protocol, bytes ->
+    requireNotNull(bytes) { "Non-null value expected for protocol $protocol" }
+    require(bytes.isNotEmpty()) { "Non-empty value expected for protocol $protocol" }
+}
+private val PEER_ID_VALIDATOR: (Protocol, ByteArray?) -> Unit = { protocol, bytes ->
+    if (bytes == null)
+        throw IllegalArgumentException("Non-null value expected for PeerId in $protocol")
+    PeerId(bytes) // constructor validates array size
+}
+
+private val NO_PARSER: (Protocol, String) -> ByteArray = { protocol, _ ->
+    throw IllegalArgumentException("No value serializer for protocol $protocol")
+}
+private val NO_DESERIALIZER: (Protocol, ByteArray) -> String = { protocol, _ ->
```

Comment:
I'm confused about the PARSER vs DESERIALIZER here.  In my mind a PARSER goes from String to internal representation and a SERIALIZER goes from internal to String.  A PARSER and DESERIALIZER would then be the same thing.

Reply:
Oh, that was also breaking my mind. 
I used to term 'serializer/deserializer' for converting 'bytes' from/to internal representation
'parser' in my mind is 'string => internal representation' converter. 
If we assume that 'bytes' is internal form, then it should be something opposite to 'parser'. 'Stringifier', 'Composer', 'Unparser'? What term do you think is the most adequate? 

Just don't want to confuse it with the 'binary serializer'. If take the `Multiaddr` class for example then it clearly has 3 representations: binary, string and internal. And they are all distinct

Reply:
In other words: what should be the name of an substance who does `toString()`? 😄

Reply:
oof 3 meaningful states makes it hard. While it's not valid English, `Stringifier` is probably a good name as it's very clear what it does.  Or possibly parser becomes `STRING_TO_BYTES` and the current deserialiser becomes `BYTES_TO_STRING`?